packages/detector.go

```diff
@@ -11,67 +11,143 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
+
+// Package packages implements package manager detection and package
+// manager specific output formatting.
 package packages
 
 import (
+	"bytes"
 	"errors"
 	"fmt"
-	"os/exec"
-	"syscall"
+	"os"
+	"regexp"
+	"strconv"
 
 	"github.com/sirupsen/logrus"
 )
 
-var execCommand = exec.Command
+// SupportedPackageManagers represents the standard error string used
+// when OS package maanger can not be identified.
+const SupportedPackageManagers = "No supported package manager found; apk, dpkg, dnf or yum installed?"
+
+// Release allows tests to inject ID and VERSION_ID typically
+// parsed from os-release.
+type Release struct {
+	ID      []byte
+	Version []byte
+}
+
+// DetectPackageManager parses os-release file to determine package
+// manager based on distribution ID.
+//
+// Optional release short circuits os-release parsing to force
+// appropriate package manager.
+func DetectPackageManager(logger *logrus.Logger, r ...Release) (string, error) {
+	var packageManager string
+	var id []byte
+	var version []byte
+
+	if len(r) == 0 {
+		data, err := readReleaseFile()
+		if err != nil {
+			return packageManager, err
+		}
+		raw := bytes.Split(data, []byte("\n"))
+
+		id, err = parseField(raw, "ID")
+		if err != nil {
+			return packageManager, err
+		}
+
+		version, err = parseField(raw, "VERSION_ID")
+		if err != nil {
+			return packageManager, err
+		}
+	} else {
+		id = r[0].ID
+		version = r[0].Version
+	}
```

Comment:
So this feels off to me. If we are going to leave it this way and pass in `Release` feels like there should be a different method that calculates the `Release` and only has one argument of logger and passes the calculated Release into this method. 

That would be used by the actual production code. And you *could* still use this method for the unit tests. But if we are being honest I think we have almost all the tests we should need here.

https://github.com/sonatype-nexus-community/ahab/tree/master/docker

We are probably missing a few versions of redhat, debian, mint, fedora, etc so we could add those to make sure this is covered for the parts of the case statement below. And if we go that way we would at least not need the extra "test code" in the production code ¯\_(ツ)_/¯

Reply:
Thanks for the sanity check! Agreed, the docker based coverage makes this redundant + it just felt wrong.

Reply:
i was going too far with the matching -- the rhel/redhat bits in particular probably don't make sense, because they are commercial (not sure how heavily used they would be as cloud images, as a former redhat user i just used centos for that). going to clean that up which will simplify the case.

i'll add more tests to cover the additional debian distros. i'm not sure if anyone would run a rolling release in the cloud, but arch also now has a docker image (finally)... honestly i'm not sure why anyone would run something other than alpine for most things, but a future PR might be adding support for arch. since that'll need a new purl format maybe that can be part of #44.

Reply:
need to think this through after more coffee tomorrow. when adding more docker based tests the first thing i hit was what looked like timeouts. boosting -timeout, i then got rate limited by OSSI. i feel like i worked around that in the past with a token, but need to remember how to get it injected into all the containers...and possibly rethink having so many parallel invocations.

Reply:
i've poked at this awhile, and the issues seems to be rate limiting -- although i might have uncovered something worth chasing.

if i run the original set of docker tests (8) all pass. if i add variants for ubuntu and old/new centos/fedora (11) each test can be ran individually but running them all errors out. playing with -timeout and -parallel doesn't help, just makes it take longer to fail. if the total number of ossi requests goes past the limit it obviously causes issues.

however, in the past when i'd hit that myself i'd see "you've been rate limited" type messages. i've been able to consistently cause the failure when running more tests, and each time i see the same message -- and it's failing (due to order) at the same test each time which is not a new variant (so unlikely it's due to something i did in the dockerfile/etc). the message is:

```console
Uh oh, an error occurred, if this persists try rerunning with -v, -vv, or -vvv to get more information in the logs
                Error: runtime error: index out of range [1] with length 1
```
it's not a lot to go on, but unless this turns out to be something silly from the changes i've made, it feels like this might be an opportunity for more defensive programming / better error message.

i'm going to try and figure out where that's coming from...but wanted to pose some questions while i'm spelunking:

- are there time limits to worry about in the current CI setup? even w/o a rate limit, more tests obviously add tiime. aside from causing longer builds, not sure if there are hard limits we need to keep in mind. it's all running with t.Parallel() but also not sure how parallelism is limited in pipelines. :thinking: 

- do the CI tests currently run in such a way that they use an account/token for OSSI access? even if this odd error is chased down, i think the additional tests will still require token access to ever succeed. i was thinking we could refactor the dockerfiles to use env vars to feed that to ahab, but that would mean the CI system needs configured house secrets and devs need .envrc or some way to consistently set the same. is this worthwhile, or should i just pair back the tests?

Reply:
sorry to spam, wanted to brain dump before i sleep and forget everything.

per gist above, i noticed it fails consistently with a dockerfile that forces centos:latest + specifying `--os fedora` (which should be using dnf). since it complains about a `[1]` index, i'm trying to figure out if it's somehow package format related since we have some stuff like this:

```console
parse/yum.go
50:     parsedProject.Version = doParseYumVersionIntoPurl(splitPackage[1])
58:     parsedProject.Version = doParseYumVersionIntoPurl(splitPackage[1])
```
i haven't dug into that code yet, but it's the only place i see a `[1]` index used in yum-related code with a quick grep.

Reply:
wow this was an edge case, i need sleep + coffee to trust the finding, but current thought is it looks like meta packages don't have versions which throws the yum parsing off. just diffing dnf list installed output before and after adding the packages which caused failures and see stuff like this after:

```console
...
+dracut-network.x86_64             049-70.git20200228.el8              @System
+dracut-squash.x86_64              049-70.git20200228.el8              @System
+dracut.x86_64                     049-70.git20200228.el8              @System
+elfutils-default-yama-scope.noarch
+elfutils-libelf.x86_64            0.178-7.el8                         @System
+elfutils-libs.x86_64              0.178-7.el8                         @System
...
```

Reply:
i've confirmed this is a legitimate issue the extended test coverage uncovered. steps to reproduce:

- docker pull centos:latest (or fedora:latest)
- docker run -it --rm -v `pwd`:`pwd` -w `pwd` centos:latest (make sure freshly built ahab binary is in pwd)
- yum (or dnf) -y install epel-release python3-pip
- yum (or dnf) list installed | ./ahab chase (or --os or --package-manager)

also verified by checking diff output and doing simple shell for/while prefixing each line -- yum/dnf list output can generate lines which miss some fields so breaks current parsing.

proposal: while working around this in various ways it dawned on me that we might avoid this issue + simplify yum vs dnf detection by simply using rpm --queryformat. rpm exists in all centos/fedora versions so simplifies detection/testing. this approach would be similar to using dpkg-query vs apt in debian distros -- apt/dnf/yum are really just dependency management tools atop the actual package manager.

any objections to reworking this to use rpm vs dnf/yum? it should ensure consistent input data by being able to control the format. for example:

```console
[root@976a05fd817a yum]# rpm -qa --queryformat "%{NAME}.%{ARCH} %{VERSION}\n" | grep elf
elfutils-libelf.x86_64 0.178
elfutils-libs.x86_64 0.178
elfutils-default-yama-scope.noarch 0.178
```